### ISSUE TITLE
Lack of dataset length or cardinality causes `BaseCollectiveExecutor::StartAbort Out of range` issues

### ISSUE THREAD

**Issue by 00krishna** [tensorflow/tensorflow#35314]
<em>Please make sure that this is a bug. As per our [GitHub Policy](https://github.com/tensorflow/tensorflow/blob/master/ISSUES.md), we only address code/doc bugs, performance issues, feature requests and build/installation issues on GitHub. tag:bug_template</em>

**System information**
- Have I written custom code (as opposed to using a stock example script provided in TensorFlow): I borrowed someone's code for this particular issue.
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): Linux Ubuntu 18.04
- Mobile device (e.g. iPhone 8, Pixel 2, Samsung Galaxy) if the issue happens on mobile device:
- TensorFlow installed from (source or binary): Source install through Conda.
- TensorFlow version (use command below): 2.0.0
- Python version: python 3.7.
- Bazel version (if compiling from source): NA.
- GCC/Compiler version (if compiling from source): Cuda 10.1
- CUDA/cuDNN version: 7.4
- GPU model and memory: Nvidia GTX 1080

You can collect some of this information using our environment capture
[script](https://github.com/tensorflow/tensorflow/tree/master/tools/tf_env_collect.sh)
You can also obtain the TensorFlow version with: 1. TF 1.0: `python -c "import
tensorflow as tf; print(tf.GIT_VERSION, tf.VERSION)"` 2. TF 2.0: `python -c
"import tensorflow as tf; print(tf.version.GIT_VERSION, tf.version.VERSION)"`

**Describe the current behavior**

There is an existing error https://github.com/tensorflow/tensorflow/issues/32817 the references 
an issue that I had raised before. The problem is that when the batch size does not divide evenly into the total number of examples, you can an error `BaseCollectiveExecutor::StartAbort Out of range.` This is actually a pretty confusing and un-intuitive error message. The code to reproduce this error is below. 

The problem is that this error message relates to a closed issue that I had raised before (https://github.com/tensorflow/tensorflow/issues/26966). There seems to be no good way to compute the cardinality or number of examples in a dataset. There is a function `tf.data.experimental.cardinality` that will find the cardinality of a dataset if say the dataset comes from a pandas dataframe or something, but generally not from some large CSV file or text file. 

Also, say I actually calculate the cardinality of the dataset by manually counting over the batches of examples. The problem is that there is no where to record this information in the Dataset API itself. So I have to either manually hardcode this in a file, or recompute the length each time I run a training loop, which takes a lot of time. 

Is there a better strategy for this? Like is there a way to encode some metadata into a tensorflow Dataset API? Or perhaps a better and more recommended way is to create a Tensorflow Dataset in the style of the https://www.tensorflow.org/datasets , i.e., `tfds`. 

**Describe the expected behavior**

I should not get a warning message for the `BaseCollectiveExecutor::StartAbort Out of range` message because the total number of examples in the dataset does not divide evenly into the batch size. Alternatively, I should be able to compute the total dataset size using the `tf.data.experimental.cardinality` function for csv files. 

**Code to reproduce the issue**
Provide a reproducible test case that is the bare minimum necessary to generate the problem.

Code taken from: https://github.com/tensorflow/tensorflow/issues/32817#issuecomment-539200561

data = tf.random.normal((60000,30,4))
ground_truth = tf.ones((60000,1))
dataset = tf.data.Dataset.from_tensor_slices((data, ground_truth)).batch(64)

#predefined model here: input: [?, 30,4] output: [?,1]
model.fit(dataset, epochs=5)

'''
    938/Unknown - 16s 17ms/step - loss: 0.02172019-10-07 14:49:49.928619: W tensorflow/core/common_runtime/base_collective_executor.cc:216] BaseCollectiveExecutor::StartAbort Out of range: End of sequence
         [[{{node IteratorGetNext}}]]
         [[Shape/_2]]
2019-10-07 14:49:49.928619: W tensorflow/core/common_runtime/base_collective_executor.cc:216] BaseCollectiveExecutor::StartAbort Out of range: End of sequence
         [[{{node IteratorGetNext}}]]
938/938 [==============================] - 16s 17ms/step - loss: 0.0217
Epoch 2/5
935/938 [============================>.] - ETA: 0s - loss: 2.2229e-062019-10-07 14:49:59.722216: W tensorflow/core/common_runtime/base_collective_executor.cc:216] BaseCollectiveExecutor::StartAbort Out of range: End of sequence
         [[{{node IteratorGetNext}}]]
2019-10-07 14:49:59.722218: W tensorflow/core/common_runtime/base_collective_executor.cc:216] BaseCollectiveExecutor::StartAbort Out of range: End of sequence
         [[{{node IteratorGetNext}}]]
         [[Shape/_2]]
'''

**Other info / logs**
Include any logs or source code that would be helpful to diagnose the problem. If including tracebacks, please include the full traceback. Large logs and files should be attached.

**Comment by jsimsa** [comment]
@dubey @omalleyt12 why is the collective executor involved here?

@rxsang @tomerk what is the recommendation for Keras users? My expectation would be that Keras can handle this scenario (`fit` with dataset whose cardinality is not divisible by batch size) correctly.

**Comment by rxsang** [comment]
@qqfish 

Hi Krishna,

It seems you are not using any DistributionStrategy, correct? Are you able to continue training even after you get the OutOfRange error log?

My suspicion is because currently the OutOfRange error is raised from C++, then propagated and caught in python. So C++ side logs are always printed regardless whether user catches the exception in python, which causes the confusion.

**Comment by dubey** [comment]
The collective executor warning probably comes from:
https://github.com/tensorflow/tensorflow/blob/bb45024ae9d3df0127d1c1056b08f25e60ba601c/tensorflow/core/common_runtime/base_collective_executor.cc#L217
which is called from:
https://github.com/tensorflow/tensorflow/blob/bb45024ae9d3df0127d1c1056b08f25e60ba601c/tensorflow/core/common_runtime/executor.cc#L2289

We can lower the logging level if it causes confusion for users.

**Comment by 00krishna** [comment]
@rxsang Sorry I did not see your message until today. I understand what you mean that the C++ warnings are propagated upward. I agree that this is a source of confusion because I could not really tell if training was happening or if some batches were being aborted or what. 

So just to clarify, should I not worry about this warning, or is there some better way to handle it? Should I adjust my own settings to lower the logging level, or is that something that has to be done in the interface between python and C++? Seems like https://github.com/tensorflow/tensorflow/commit/4ab663e27c230a99e2e67f0609932d6151c54a32 should have fixed this, so would I just need to upgrade my installation, or do I have to wait for the next release. 

Thanks so much for your help. Yeah, running the counter over all the batches was certainly taking a lot of time. So if I don't have to do that, it will be great. 
Krishna

**Comment by dubey** [comment]
Yes, you can ignore the warning.  The fix I submitted is available in nightly and should be a part of the next release.

**Comment by 00krishna** [comment]
Thanks so much @dubey

**Comment by durandg12** [comment]
> Yes, you can ignore the warning. The fix I submitted is available in nightly and should be a part of the next release.

I can confirm that the warning disappeared for me when using tf-nightly.
